Content.Client/UserInterface/Systems/Ghost/Controls/Roles/GhostRolesEui.cs

```diff
@@ -74,10 +103,29 @@ public override void HandleState(EuiStateBase state)
                 _window.AddEntry(name, description, group);
             }
 
-            var closeRulesWindow = ghostState.GhostRoles.All(role => role.Identifier != _windowRulesId);
-            if (closeRulesWindow)
+            if (_windowRules != null && !_windowRules.RoleHasGone)
             {
-                _windowRules?.Close();
+                // Perhaps the user won't be able to get this role now.
+                _windowRules.RoleHasGone = ghostState.GhostRoles.All(role => role.Identifier != _windowRulesId);
+
+                if (_windowRules.RoleHasGone)
+                {
+                    var replacements =
+                        ghostState.GhostRoles.Where(role => role.Name == _windowRulesName).Select(role=> role.Identifier).ToList();
+
+                    if (replacements.Count == 1)
+                    {
+                        _windowRulesId = replacements[0];
+                        _windowRules.RoleHasGone = false;
+                    }
+                    else if (replacements.Count >= 2)
+                    {
+                        // Pick randomly or else multiple queued players will all drop until the first element
+                        var index = _random.Next(replacements.Count - 1);
+                        _windowRulesId = replacements[index];
+                        _windowRules.RoleHasGone = false;
+                    }
+                }
             }
```

Comment:
This needs to disclose when failure occurs as someone might mash the button if it takes more than a second due to ping or the ghost role legitimately being taken already.